File: scripts/topics.py
"""AutoPub F1 — Motor de ideas (CLI keyless).

Genera briefs de contenido priorizados desde fuentes RSS + búsqueda DuckDuckGo
(solo stdlib: urllib + xml.etree + json), con el MISMO esquema de brief que la
tool TS `packages/core/src/tools/topics.ts` (fuente de verdad):

    {tema, canal, formato, tono, angulo, fuentes, score, pubDate}

Uso:
    python scripts/topics.py --dry-run                 # briefs a stdout (default)
    python scripts/topics.py --dry-run --max 12        # más briefs
    python scripts/topics.py --canales yt,tiktok       # solo canales elegidos
    python scripts/topics.py --out briefs.json         # escribe JSON (UTF-8 sin BOM)
    python scripts/topics.py --rss <url> --search "IA" # fuentes custom (repetible)

Reglas aprendidas aplicadas: API directa > scraping; JSON sin BOM (PowerShell);
degradación elegante si una fuente falla (nunca crashea).
"""
from __future__ import annotations

import argparse
import json
import re
import sys
import urllib.request
import xml.etree.ElementTree as ET
from datetime import datetime, timezone
from typing import Any

HEADERS = {"User-Agent": "Mozilla/5.0 (UltraIa topics/0.1)"}

DEFAULT_RSS = [
    "https://hnrss.org/frontpage",
    "https://feeds.arstechnica.com/arstechnica/technology-lab",
    "https://www.theverge.com/rss/index.xml",
]

DEFAULT_SEARCHES = ["AI tendencias 2026", "IA generativa noticias hoy"]

CANALES = {
    "youtube_shorts": {"formato": "9:16 video", "keywords": ["tutorial", "como", "tips", "5", "mejores", "error", "rapido", "facil"]},
    "tiktok": {"formato": "9:16 video", "keywords": ["tendencia", "viral", "hack", "lifehack", "misterio", "antes", "despues", "pov"]},
    "instagram": {"formato": "1:1 imagen", "keywords": ["estetica", "diseno", "inspiracion", "idea", "pack", "branding", "visual"]},
    "blog": {"formato": "16:9 articulo", "keywords": ["guia", "analisis", "futuro", "estrategia", "que es", "como funciona", "reporte", "caso"]},
}


def normalize_title(title: str) -> str:
    return re.sub(r"\s+", " ", re.sub(r"[^\w\s]", " ", title.lower())).strip()


def bigrams(s: str) -> set[str]:
    tokens = [t for t in s.split(" ") if t]
    return {f"{tokens[i]} {tokens[i + 1]}" for i in range(len(tokens) - 1)}


def jaccard(a: set[str], b: set[str]) -> float:
    if not a and not b:
        return 1.0
    inter = len(a & b)
    union = len(a | b)
    return inter / union if union else 0.0


def dedupe(items: list[dict[str, Any]]) -> list[dict[str, Any]]:
    seen: list[str] = []
    out: list[dict[str, Any]] = []
    for item in items:
        norm = normalize_title(item["title"])
        if not norm:
            continue
        if any(jaccard(bigrams(prev), bigrams(norm)) > 0.6 for prev in seen):
            continue
        seen.append(norm)
        out.append(item)
    return out


def fetch(url: str, timeout: int = 20) -> str:
    req = urllib.request.Request(url, headers=HEADERS)
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        return resp.read().decode("utf-8", errors="replace")


def parse_rss(url: str, max_items: int = 8) -> list[dict[str, Any]]:
    xml_text = fetch(url)
    root = ET.fromstring(xml_text)
    items: list[dict[str, Any]] = []
    for node in root.iter("item"):
        if len(items) >= max_items:
            break
        title = (node.findtext("title") or "").strip()
        link = (node.findtext("link") or url).strip()
        desc = (node.findtext("description") or "").strip()
        desc = re.sub(r"<[^>]+>", " ", desc)
        pub = (node.findtext("pubDate") or "").strip()
        items.append({"title": title, "link": link, "description": desc, "pubDate": pub or None})
    return items


def ddg_search(query: str, max_results: int = 8) -> list[dict[str, Any]]:
    api = f"https://api.duckduckgo.com/?q={urllib.parse.quote(query)}&format=json&no_html=1&skip_disambig=1"
    data = json.loads(fetch(api))
    items: list[dict[str, Any]] = []
    if data.get("AbstractText") and data.get("AbstractURL"):
        items.append({"title": data.get("Heading") or data["AbstractURL"], "link": data["AbstractURL"], "description": data["AbstractText"], "pubDate": None})
    for topic in data.get("RelatedTopics") or []:
        if len(items) >= max_results:
            break
        if isinstance(topic, dict) and topic.get("Text") and topic.get("FirstURL"):
            text = str(topic["Text"])
            items.append({"title": text.split(" - ")[0], "link": topic["FirstURL"], "description": text, "pubDate": None})
    return items[:max_results]


def parse_date(pub: str | None) -> datetime | None:
    if not pub:
        return None
    # RFC 2822 (RSS). Si falla, devuelve None (score neutro).
    try:
        from email.utils import parsedate_to_datetime

        return parsedate_to_datetime(pub).astimezone(timezone.utc)
    except Exception:
        return None


def novelty(pub: str | None, now: datetime) -> float:
    dt = parse_date(pub)
    if dt is None:
        return 0.5
    age_days = (now - dt).total_seconds() / 86400
    if age_days < 0 or age_days <= 7:
        return 1.0
    if age_days >= 30:
        return 0.0
    return 1.0 - (age_days - 7) / 23


def tone_for(title: str, desc: str, from_search: bool) -> str:
    t = f"{title} {desc}".lower()
    if re.search(r"\b(guia|como|tutorial|paso|tips|hack|mejores)\b", t):
        return "educativo"
    if re.search(r"\b(tendencia|viral|misterio|pov|antes|despues)\b", t):
        return "entretenido"
    if re.search(r"\b(analisis|reporte|futuro|estrategia|estudio|datos)\b", t):
        return "analitico"
    if re.search(r"\b(nueva|nuevo|anuncio|lanza|presenta)\b", t):
        return "noticia"
    if re.search(r"\b(inspira|idea|pack|branding|estetica|diseno)\b", t):
        return "inspirador"
    return "informativo" if from_search else "noticia"


def angle_for(title: str, desc: str, from_search: bool) -> str:
    t = f"{title} {desc}".lower()
    if re.search(r"\b(como|tutorial|paso)\b", t):
        return "Tutorial paso a paso con ejemplo real"
    if re.search(r"\b(tips|hack|mejores)\b", t):
        return "Lista de recomendaciones accionables"
    if re.search(r"\b(tendencia|viral)\b", t):
        return "Por que esto esta en tendencia ahora"
    if re.search(r"\b(guia|estrategia|analisis)\b", t):
        return "Analisis con contexto y datos"
    if re.search(r"\b(nueva|nuevo|anuncio|lanza)\b", t):
        return "Novedad explicada en 60 segundos"
    return "Resumen de tendencia con datos de la busqueda" if from_search else "Resumen de la noticia con contexto"


def build_briefs(items: list[dict[str, Any]], canales: list[str], now: datetime) -> list[dict[str, Any]]:
    briefs: list[dict[str, Any]] = []
    for item in items:
        from_search = item.get("from_search", False)
        for canal in canales:
            cfg = CANALES[canal]
            norm = normalize_title(item["title"])
            relevance = 1.0 if any(kw in norm for kw in cfg["keywords"]) else 0.5
            score = round(novelty(item.get("pubDate"), now) * relevance, 2)
            briefs.append(
                {
                    "tema": item["title"],
                    "canal": canal,
                    "formato": cfg["formato"],
                    "tono": tone_for(item["title"], item["description"], from_search),
                    "angulo": angle_for(item["title"], item["description"], from_search),
                    "fuentes": [item["link"]],
                    "score": score,
                    "pubDate": item.get("pubDate"),
                }
            )
    briefs.sort(key=lambda b: b["score"], reverse=True)
    return briefs


def main() -> int:
    import urllib.parse  # noqa: PLC0415 (import tardío para CLI corto)

    parser = argparse.ArgumentParser(description="AutoPub F1 — motor de ideas keyless (RSS + DuckDuckGo)")
    parser.add_argument("--dry-run", action="store_true", help="imprime briefs sin persistir (default)")
    parser.add_argument("--max", type=int, default=8, help="máximo de briefs (default 8)")
    parser.add_argument("--canales", default="youtube_shorts,tiktok,instagram,blog", help="canales separados por coma")
    parser.add_argument("--rss", action="append", default=[], help="feed RSS custom (repetible)")
    parser.add_argument("--search", action="append", default=[], help="query de tendencias (repetible)")
    parser.add_argument("--out", default=None, help="ruta del JSON de salida (UTF-8 sin BOM)")
    args = parser.parse_args()

    rss_sources = args.rss or DEFAULT_RSS
    search_sources = args.search or DEFAULT_SEARCHES
    canales = [c.strip() for c in args.canales.split(",") if c.strip() in CANALES]

    now = datetime.now(timezone.utc)
    raw: list[dict[str, Any]] = []
    activas: list[str] = []

    for url in rss_sources:
        try:
            items = parse_rss(url)
            if items:
                activas.append(url)
                raw.extend(items)
        except Exception as exc:  # noqa: BLE001 — degradación elegante por fuente
            print(f"[topics] aviso: RSS {url} falló ({exc}) — se ignora", file=sys.stderr)

    for q in search_sources:
        try:
            items = ddg_search(q)
            if items:
                activas.append(f"search:{q}")
                for it in items:
                    it["from_search"] = True
                raw.extend(items)
        except Exception as exc:  # noqa: BLE001
            print(f"[topics] aviso: búsqueda '{q}' falló ({exc}) — se ignora", file=sys.stderr)

    unique = dedupe(raw)
    briefs = build_briefs(unique, canales, now)[: args.max]

    result = {
        "generatedAt": now.isoformat(),
        "rawCount": len(raw),
        "uniqueCount": len(unique),
        "fuentesActivas": activas,
        "briefs": briefs,
    }

    if args.out:
        with open(args.out, "w", encoding="utf-8", newline="\n") as fh:
            json.dump(result, fh, ensure_ascii=False, indent=2)
        print(f"[topics] {len(briefs)} briefs escritos en {args.out}")
    else:
        print(json.dumps(result, ensure_ascii=False, indent=2))

    return 0 if briefs else 1


if __name__ == "__main__":
    sys.exit(main())

File: packages/core/src/tools/topics.ts
/**
 * AutoPub F1 — Motor de ideas (topics).
 *
 * Genera briefs de contenido recurrentes sin intervención manual, keyless-first:
 *   - Fuentes: RSS feeds (parseRss de reach) + búsqueda web DuckDuckGo (tendencias).
 *   - Dedupe por título normalizado + bigram overlap.
 *   - Priorización: score = novedad (frescura del pubDate) × relevancia (keywords del canal).
 *   - Brief JSON estandarizado: {tema, canal, formato, tono, angulo, fuentes, score}.
 *
 * Diseño: `fetchFn` inyectable (para tests sin red); por defecto usa globalThis.fetch.
 * La tool TS es la fuente de verdad; `scripts/topics.py` es el CLI autónomo keyless con el
 * mismo esquema de brief (documentado en AUTO-PUBLICACION.md §4 F1).
 */

import { parseRss, searchWeb } from './reach';

/** Canal objetivo de publicación (D1/D8 del plan maestro; telegram/discord/slack = canales de mensajería; facebook = canal video, iteraciones 39-40, 53). */
export type TopicChannel = 'youtube_shorts' | 'tiktok' | 'instagram' | 'blog' | 'telegram' | 'discord' | 'slack' | 'facebook';

/** Formato visual que el canal espera (F3 `present` lo consumirá). */
export type TopicFormat = '9:16 video' | '1:1 imagen' | '16:9 articulo' | '16:9 video';

/** Tono editorial sugerido para el brief. */
export type TopicTone = 'informativo' | 'educativo' | 'entretenido' | 'inspirador' | 'analitico' | 'noticia';

export interface TopicSource {
  /** URL de un feed RSS/Atom. */
  rss?: string;
  /** Query de búsqueda de tendencias (DuckDuckGo, keyless). */
  search?: string;
}

export interface TopicBrief {
  tema: string;
  canal: TopicChannel;
  formato: TopicFormat;
  tono: TopicTone;
  angulo: string;
  fuentes: string[];
  score: number;
  pubDate: string | null;
}

export interface GenerateTopicsInput {
  /** Fuentes de temas (RSS y/o búsquedas). Vacío → usa las fuentes por defecto. */
  fuentes?: TopicSource[];
  /** Canales objetivo (default: los 4). */
  canales?: TopicChannel[];
  /** Máximo de briefs a devolver (default 8). */
  maxBriefs?: number;
}

export interface GenerateTopicsResult {
  briefs: TopicBrief[];
  /** Cuántos ítems crudos se vieron (pre-dedupe) por fuente. */
  rawCount: number;
  /** Cuántos ítems únicos tras dedupe. */
  uniqueCount: number;
  /** Fuentes que respondieron (RSS con items o búsqueda con resultados). */
  fuentesActivas: string[];
}

/** Keywords por canal — ponderan la relevancia en el score. */
const CHANNEL_KEYWORDS: Record<TopicChannel, string[]> = {
  youtube_shorts: ['tutorial', 'como', 'tips', '5', 'mejores', 'error', 'rapido', 'facil'],
  tiktok: ['tendencia', 'viral', 'hack', 'lifehack', 'misterio', 'antes', 'despues', 'pov'],
  instagram: ['estetica', 'diseno', 'inspiracion', 'idea', 'pack', 'branding', 'visual'],
  blog: ['guia', 'analisis', 'futuro', 'estrategia', 'que es', 'como funciona', 'reporte', 'caso'],
  telegram: ['canal', 'diario', 'resumen', 'noticia', 'actualizacion', 'inteligencia artificial'],
  discord: ['comunidad', 'canal', 'noticia', 'anuncio', 'debate', 'resumen'],
  slack: ['equipo', 'actualizacion', 'resumen', 'anuncio', 'interno', 'reporte'],
  facebook: ['reels', 'video', 'tendencia', 'compartir', 'viral', 'comunidad', 'pagina'],
};

/** Mapa canal → formato visual (F3 `present` lo usará). */
const FORMAT_BY_CHANNEL: Record<TopicChannel, TopicFormat> = {
  youtube_shorts: '9:16 video',
  tiktok: '9:16 video',
  instagram: '1:1 imagen',
  blog: '16:9 articulo',
  telegram: '9:16 video',
  discord: '9:16 video',
  slack: '9:16 video',
  facebook: '9:16 video',
};

/** Tono por defecto según el origen del ítem. */
function toneFor(item: { title: string; description: string }, fromSearch: boolean): TopicTone {
  const t = `${item.title} ${item.description}`.toLowerCase();
  if (/\b(guia|como|tutorial|paso|tips|hack|mejores)\b/.test(t)) return 'educativo';
  if (/\b(tendencia|viral|misterio|pov|antes|despues)\b/.test(t)) return 'entretenido';
  if (/\b(analisis|reporte|futuro|estrategia|estudio|datos)\b/.test(t)) return 'analitico';
  if (/\b(nueva|nuevo|anuncio|lanza|presenta)\b/.test(t)) return 'noticia';
  if (/\b(inspira|idea|pack|branding|estetica|diseno)\b/.test(t)) return 'inspirador';
  return fromSearch ? 'informativo' : 'noticia';
}

/** Ángulo sugerido (template por tipo de fuente). */
function angleFor(item: { title: string; description: string }, fromSearch: boolean): string {
  const t = `${item.title} ${item.description}`.toLowerCase();
  if (/\b(como|tutorial|paso)\b/.test(t)) return 'Tutorial paso a paso con ejemplo real';
  if (/\b(tips|hack|mejores)\b/.test(t)) return 'Lista de recomendaciones accionables';
  if (/\b(tendencia|viral)\b/.test(t)) return 'Por que esto esta en tendencia ahora';
  if (/\b(guia|estrategia|analisis)\b/.test(t)) return 'Analisis con contexto y datos';
  if (/\b(nueva|nuevo|anuncio|lanza)\b/.test(t)) return 'Novedad explicada en 60 segundos';
  return fromSearch ? 'Resumen de tendencia con datos de la busqueda' : 'Resumen de la noticia con contexto';
}

/** Normaliza un título: minúsculas + sin puntuación + espacios colapsados. */
export function normalizeTitle(title: string): string {
  return title
    .toLowerCase()
    .replace(/[^\p{L}\p{N}\s]/gu, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

/** Bigrams de un texto normalizado. */
function bigrams(s: string): Set<string> {
  const tokens = s.split(' ').filter(Boolean);
  const out = new Set<string>();
  for (let i = 0; i < tokens.length - 1; i++) out.add(`${tokens[i]} ${tokens[i + 1]}`);
  return out;
}

/** Similitud de Jaccard entre bigram sets (0..1). */
function jaccard(a: Set<string>, b: Set<string>): number {
  if (!a.size && !b.size) return 1;
  let inter = 0;
  for (const x of a) if (b.has(x)) inter++;
  const union = a.size + b.size - inter;
  return union === 0 ? 0 : inter / union;
}

/** Dedupe: agrupa ítems por similitud de título (Jaccard bigram > 0.6). */
export function dedupeItems<T extends { title: string }>(items: T[]): T[] {
  const seen: string[] = [];
  const out: T[] = [];
  for (const item of items) {
    const norm = normalizeTitle(item.title);
    if (!norm) continue;
    const dup = seen.some((prev) => jaccard(bigrams(prev), bigrams(norm)) > 0.6);
    if (dup) continue;
    seen.push(norm);
    out.push(item);
  }
  return out;
}

/** Peso de novedad: 0..1 según antigüedad del pubDate (7 días → 1, 30+ días → 0). */
export function noveltyScore(pubDate: string | null, now: number = Date.now()): number {
  if (!pubDate) return 0.5;
  const t = Date.parse(pubDate);
  if (Number.isNaN(t)) return 0.5;
  const ageDays = (now - t) / 86_400_000;
  if (ageDays < 0) return 1;
  if (ageDays <= 7) return 1;
  if (ageDays >= 30) return 0;
  return 1 - (ageDays - 7) / 23;
}

/** Score final: novedad × relevancia de keywords del canal × bonus de cobertura. */
export function scoreBrief(
  brief: { pubDate: string | null; tema: string },
  canal: TopicChannel,
  now: number = Date.now(),
): number {
  const novelty = noveltyScore(brief.pubDate, now);
  const t = normalizeTitle(brief.tema);
  let relevance = 0.5;
  const channelKw = CHANNEL_KEYWORDS[canal];
  if (channelKw && channelKw.some((kw) => t.includes(kw))) relevance = 1;
  return Math.round(novelty * relevance * 100) / 100;
}

/** Fuentes por defecto cuando el input no trae ninguna (RSS tech + tendencias DDG). */
export const DEFAULT_SOURCES: TopicSource[] = [
  { rss: 'https://hnrss.org/frontpage' },
  { rss: 'https://feeds.arstechnica.com/arstechnica/technology-lab' },
  { rss: 'https://www.theverge.com/rss/index.xml' },
  { search: 'AI tendencias 2026' },
  { search: 'IA generativa noticias hoy' },
];

interface RawItem {
  title: string;
  link: string;
  description: string;
  pubDate: string | null;
  fuente: string;
  fromSearch: boolean;
}

/** Recoge ítems crudos de las fuentes activas (RSS + búsqueda). */
async function collectRaw(input: GenerateTopicsInput): Promise<{ items: RawItem[]; activas: string[] }> {
  const fuentes = input.fuentes?.length ? input.fuentes : DEFAULT_SOURCES;
  const items: RawItem[] = [];
  const activas: string[] = [];

  const tasks = fuentes.map(async (src) => {
    if (src.rss) {
      try {
        const res = await parseRss({ url: src.rss, maxItems: 8 });
        if (res.items.length) {
          activas.push(src.rss);
          for (const it of res.items) {
            items.push({
              title: it.title || '(sin titulo)',
              link: it.link || src.rss,
              description: it.description,
              pubDate: it.pubDate,
              fuente: res.feedTitle || src.rss,
              fromSearch: false,
            });
          }
        }
      } catch {
        // fuente muerta → se ignora (degradación elegante)
      }
    }
    if (src.search) {
      try {
        const res = await searchWeb({ query: src.search, maxResults: 8 });
        if (res.results.length) {
          activas.push(`search:${src.search}`);
          for (const r of res.results) {
            items.push({
              title: r.title || '(sin titulo)',
              link: r.url,
              description: r.snippet,
              pubDate: null,
              fuente: `search:${src.search}`,
              fromSearch: true,
            });
          }
        }
      } catch {
        // búsqueda fallida → se ignora
      }
    }
  });
  await Promise.all(tasks);
  return { items, activas };
}

/** Genera briefs priorizados a partir de fuentes RSS + búsqueda de tendencias. */
export async function generateTopicBriefs(input: GenerateTopicsInput = {}): Promise<GenerateTopicsResult> {
  const canales: TopicChannel[] = input.canales?.length ? input.canales : ['youtube_shorts', 'tiktok', 'instagram', 'blog'];
  const maxBriefs = Math.max(1, input.maxBriefs ?? 8);

  const { items, activas } = await collectRaw(input);
  const unique = dedupeItems(items);
  const briefs: TopicBrief[] = [];

  for (const item of unique) {
    // Cada ítem genera un brief por canal objetivo (rotando el ángulo para evitar duplicados).
    for (const canal of canales) {
      const brief: TopicBrief = {
        tema: item.title,
        canal,
        formato: FORMAT_BY_CHANNEL[canal],
        tono: toneFor(item, item.fromSearch),
        angulo: angleFor(item, item.fromSearch),
        fuentes: [item.link, item.fuente].filter(Boolean),
        score: 0,
        pubDate: item.pubDate,
      };
      brief.score = scoreBrief(brief, canal);
      briefs.push(brief);
    }
  }

  briefs.sort((a, b) => b.score - a.score);
  return { briefs: briefs.slice(0, maxBriefs), rawCount: items.length, uniqueCount: unique.length, fuentesActivas: activas };
}

export const topics = { generateTopicBriefs, normalizeTitle, dedupeItems, noveltyScore, scoreBrief };